plugin/pkg/scheduler/core/generic_scheduler.go

```diff
@@ -424,6 +445,74 @@ func EqualPriorityMap(_ *v1.Pod, _ interface{}, nodeInfo *schedulercache.NodeInf
 	}, nil
 }
 
+// selectNodesForPreemption finds all the nodes with possible victims for
+// preemption in parallel.
+func selectNodesForPreemption(pod *v1.Pod,
+	nodeNameToInfo map[string]*schedulercache.NodeInfo,
+	nodes []*v1.Node,
+	predicates map[string]algorithm.FitPredicate,
+	metadataProducer algorithm.MetadataProducer,
+) map[string][]*v1.Pod {
+
+	nodeNameToPods := map[string][]*v1.Pod{}
+	var resultLock sync.Mutex
+
+	// We can use the same metadata producer for all nodes.
+	meta := metadataProducer(pod, nodeNameToInfo)
+	checkNode := func(i int) {
+		nodeName := nodes[i].Name
+		pods, fits := selectVictimsOnNode(pod, meta, nodeNameToInfo[nodeName], predicates)
+		if fits {
+			resultLock.Lock()
+			nodeNameToPods[nodeName] = pods
+			resultLock.Unlock()
+		}
+	}
+	workqueue.Parallelize(16, len(nodes), checkNode)
+	return nodeNameToPods
+}
+
+// selectVictimsOnNode finds minimum set of pods on the given node that should
+// be preempted in order to make enough room for "pod" to be scheduled.
+// The algorithm first checks if the pod can be scheduled on the node when all the
+// lower priority pods are gone. If so, it sorts all the lower priority pods by
+// their priority and starting from the highest priority one, tries to keep as
+// many of them as possible while checking that the "pod" can still fit on the node.
+func selectVictimsOnNode(pod *v1.Pod, meta interface{}, nodeInfo *schedulercache.NodeInfo, predicates map[string]algorithm.FitPredicate) ([]*v1.Pod, bool) {
+	victims := []*v1.Pod{}
+	higherPriority := func(pod1, pod2 interface{}) bool {
+		return util.GetPodPriority(pod1.(*v1.Pod)) > util.GetPodPriority(pod2.(*v1.Pod))
+	}
+	podList := util.SortableList{CompFunc: higherPriority}
+	nodeInfoCopy := nodeInfo.Clone()
+	// As the first step, remove all the lower priority pods from the node and
+	// check if the given pod can be scheduled.
+	podPriority := util.GetPodPriority(pod)
+	for _, p := range nodeInfoCopy.Pods() {
+		if util.GetPodPriority(p) < podPriority {
+			podList.Items = append(podList.Items, p)
+			nodeInfoCopy.RemovePod(p)
+		}
+	}
+	// If the new pod does not fit after removing all the lower priority pods,
+	// we are done and this node is not suitable for preemption.
+	if fits, _, _ := podFitsOnNode(pod, meta, nodeInfoCopy, predicates, nil); !fits {
+		return nil, false
+	}
+	// If the new pod fits after the removal of all the lower priority pods, try to
+	// reprieve as may pods as possible starting from the highest priority one.
+	podList.Sort()
+	for _, p := range podList.Items {
+		lpp := p.(*v1.Pod)
+		nodeInfoCopy.AddPod(lpp)
+		if fits, _, _ := podFitsOnNode(pod, meta, nodeInfoCopy, predicates, nil); !fits {
+			nodeInfoCopy.RemovePod(lpp)
```

Comment:
If only check it per node, not sure whether it will make preempted pod preempt other lower priority; maybe return all low priority pods. Anyway, let me think about it :).